Mathlib/Control/Basic.lean

```diff
@@ -0,0 +1,269 @@
+theorem fish_pipe {α β} (f : α → m β) : pure >=> f = f := by simp only [(· >=> ·), functor_norm]
+#align fish_pipe fish_pipe
+
+-- note: in Lean 3 `>=>` is left-associative, but in Lean 4 it is right-associative.
+@[functor_norm]
+theorem fish_assoc {α β γ φ} (f : α → m β) (g : β → m γ) (h : γ → m φ) :
+    (f >=> g) >=> h = f >=> g >=> h := by
+  simp only [(· >=> ·), functor_norm]
+#align fish_assoc fish_assoc
+
+variable {β' γ' : Type v}
+
+variable {m' : Type v → Type w} [Monad m']
+
+/-- Takes a value `β` and `List α` and accumulates pairs according to a monadic function `f`.
+Accumulation occurs from the right (i.e., starting from the tail of the list). -/
+def List.mapMAccumR (f : α → β' → m' (β' × γ')) : β' → List α → m' (β' × List γ')
+  | a, [] => pure (a, [])
+  | a, x :: xs => do
+    let (a', ys) ← List.mapMAccumR f a xs
+    let (a'', y) ← f x a'
+    pure (a'', y :: ys)
+#align list.mmap_accumr List.mapMAccumR
+
+/-- Takes a value `β` and `List α` and accumulates pairs according to a monadic function `f`.
+Accumulation occurs from the left (i.e., starting from the head of the list). -/
+def List.mapMAccumL (f : β' → α → m' (β' × γ')) : β' → List α → m' (β' × List γ')
+  | a, [] => pure (a, [])
+  | a, x :: xs => do
+    let (a', y) ← f a x
+    let (a'', ys) ← List.mapMAccumL f a' xs
+    pure (a'', y :: ys)
+#align list.mmap_accuml List.mapMAccumL
+
+end Monad
+
+section
+
+variable {m : Type u → Type u} [Monad m] [LawfulMonad m]
+
+theorem joinM_map_map {α β : Type u} (f : α → β) (a : m (m α)) :
+  joinM (Functor.map f <$> a) = f <$> joinM a := by
+  simp only [joinM, (· ∘ ·), id.def, ← bind_pure_comp, bind_assoc, map_bind, pure_bind]
+#align mjoin_map_map joinM_map_map
+
+theorem joinM_map_joinM {α : Type u} (a : m (m (m α))) : joinM (joinM <$> a) = joinM (joinM a) := by
+  simp only [joinM, (· ∘ ·), id.def, map_bind, ← bind_pure_comp, bind_assoc, pure_bind]
+#align mjoin_map_mjoin joinM_map_joinM
+
+@[simp]
+theorem joinM_map_pure {α : Type u} (a : m α) : joinM (pure <$> a) = a := by
+  simp only [joinM, (· ∘ ·), id.def, map_bind, ← bind_pure_comp, bind_assoc, pure_bind, bind_pure]
+#align mjoin_map_pure joinM_map_pure
+
+@[simp]
+theorem joinM_pure {α : Type u} (a : m α) : joinM (pure a) = a :=
+  LawfulMonad.pure_bind a id
+#align mjoin_pure joinM_pure
+
+end
+
+section Alternative
+
+variable {F : Type → Type v} [Alternative F]
+
+-- [todo] add notation for `Functor.mapConst` and port `functor.map_const_rev`
+/-- Returns `pure true` if the computatoin succeeds and `pure false` otherwise. -/
+def succeeds {α} (x : F α) : F Bool :=
+  Functor.mapConst true x <|> pure false
+#align succeeds succeeds
+
+/-- Attempts to perform the computation, but fails silently if it doesn't succeed. -/
+def tryM {α} (x : F α) : F Unit :=
+  Functor.mapConst () x <|> pure ()
+#align mtry tryM
+
+@[simp]
+theorem guard_true {h : Decidable True} : @guard F _ True h = pure () := by simp [guard, if_pos]
+#align guard_true guard_True
```

Comment:
Is it `guard_true` or `guard_True`? Same for false

Reply:
The new align precheck is really nice for catching these. It's `guard_true` because definitions (or types) which appear in theorem names get `lowerCamelCased`.